heat/optim/dp_optimizer.py

```diff
@@ -1,6 +1,788 @@
+import inspect
+import math
 import torch
+import torch.distributed
+from torch.nn.parallel import DistributedDataParallel as tDDP
+from typing import Union, List, Tuple, Dict
 
-__all__ = ["DataParallelOptimizer"]
+from ..core.communication import MPICommunication
+from ..core.communication import MPI
+from ..core.communication import MPI_WORLD
+from .utils import DetectMetricPlateau
+
+
+__all__ = ["DataParallelOptimizer", "DASO"]
+
+
+def __sum_f16_cb(buffer_a, buffer_b, _):
+    # MPI custom sum function to use torch.half
+    tens_a = torch.HalfTensor().set_(torch.HalfStorage.from_buffer(buffer_a, "native"))
+    tens_b = torch.HalfTensor().set_(torch.HalfStorage.from_buffer(buffer_b, "native"))
+    tens_b += tens_a
+    nelem = torch.prod(torch.tensor(tens_b.shape)).item()
+    new_buff = MPI.memory.fromaddress(tens_b.data_ptr(), nbytes=tens_b.element_size() * nelem)
+    buffer_b[:] = new_buff
+
+
+def __sum_bfloat_cb(buffer_a, buffer_b, _):
+    # MPI custom sum function to use torch.bfloat16
+    tens_a = torch.BFloat16Tensor().set_(torch.BFloat16Storage.from_buffer(buffer_a, "native"))
+    tens_b = torch.BFloat16Tensor().set_(torch.BFloat16Storage.from_buffer(buffer_b, "native"))
+    tens_b += tens_a
+    nelem = int(tens_b.numel())
+    new_buff = MPI.memory.fromaddress(tens_b.data_ptr(), nbytes=nelem * tens_b.element_size())
+    buffer_b[:] = new_buff
+
+
+# create new MPI OPs
+mpi_sum_f16 = MPI.Op.Create(__sum_f16_cb, commute=True)
+mpi_sum_bfloat = MPI.Op.Create(__sum_bfloat_cb, commute=True)
+
+
+class DASO:
```

Comment:
Type annotations sometimes available, sometimes not. Is there a reason for this?

Reply:
nothing aside from documenting at different times. will correct now